### ISSUE TITLE
SSL3_ALIGN_PAYLOAD calculation goes wrong during do_ssl3_write

### ISSUE THREAD

**Issue by nageshshegde** [openssl/openssl#11766]
OpenSSL version 1.1.1d

During SSL_write, it is seen that the write buffer is allocated in ssl3_setup_write_buffer. When the buffer is allocated, it is considered that the align payload is of 3 bytes, and a total of 16472 bytes are allocated (16384 data + 80 (SSL3_RT_SEND_MAX_ENCRYPTED_OVERHEAD) +5 (headerlen)+align). 

When the data is copied to the buffer in do_ssl3_write, the alignment is calculated again. Here, the alignment calculation is based on the address that is allocated for the send buffer. If the memory address of the allocated buffer ends with 0x4 or 0xC, then the calculated align will be of 7bytes in do_ssl3_write function and if the address of the buffer ends with 0x0 or 0x8, then the align will be of 3bytes. 
 
In case, if the data to be sent is of 16384 bytes and if the align is 7 bytes, then the write fails in do_ssl3_write function as the data cannot be fit in the buffer of size 16472 bytes (16384+ 5(header) + 80(SSL3_RT_SEND_MAX_ENCRYPTED_OVERHEAD) +7 bytes align).

Thanks,
Nagesh

**Comment by mattcaswell** [comment]
It does seem that you are correct that the alignment calculation could be off by up to 4 bytes. Fortunately, I don't think in practice this would ever result in the buffer overrunning because `SSL3_RT_SEND_MAX_ENCRYPTED_OVERHEAD` is actually an overestimate. It consists of two parts:

````
# define SSL3_RT_SEND_MAX_ENCRYPTED_OVERHEAD \
                        (SSL_RT_MAX_CIPHER_BLOCK_SIZE + SSL3_RT_MAX_MD_SIZE)
````
Where `SSL3_RT_MAX_MD_SIZE` is defined as follows:

````
/*
 * This is the maximum MAC (digest) size used by the SSL library. Currently
 * maximum of 20 is used by SHA1, but we reserve for future extension for
 * 512-bit hashes.
 */

# define SSL3_RT_MAX_MD_SIZE                     64
````
The maximum MAC size generated by any of our ciphersuites is actually HMAC-384 == 48 bytes - so we have a headroom of 16 bytes, i.e. more than the 4 that we could overrun due to the alignment error.

**Comment by mattcaswell** [comment]
My proposed fix is in #11768.
